### DISCUSSION TITLE
RedissonMultiLock that operates in "bulk"?

### DISCUSSION THREAD

**Discussion by alexbzmn** [redisson/redisson#4630] [Q&A]
Hello everyone, Niko

As far as I understand, currently RedissonMultiLock would act as a wrapper over individual RLock(s) and therefore under a common usage pattern (distributed lock et al.) for each RLock within a multi lock, we would perform a few network calls (to redis) to lock/unlock/check. This can be quite inefficient given network calls are slow. Using atomic operations over redisson collections (`tryAdd` in RSet) to simulate "lock-like" behaviour in bulk is cumbersome as one needs to implement a watchdog, looses granular statistics and tools over regular redis keys, and is unable to provide reentrancy properties. 

Is there a strong reason against implementing a multi lock that would perform all operations in "bulk"? E.g. executing a for loop on the redis server (lua EVAL) and performing lock/unlock operations for the provided keys, the same way as it is currently done for individual RLocks.

Thanks,
Alex

**Comment by borisdainson** [comment]
I have requests that require taking about a thousand individual locks so "bulk" implementation will indeed be very helpful

**Comment by mrniko** [comment]
@alexbzmn 

Hi,

Which Redis setup do you use?

**Reply by alexbzmn** [reply]
Hello,
Sentinel setup
